### DISCUSSION TITLE
INSPIRESat-1

### DISCUSSION THREAD

**Discussion by Dhruva-Ananth** [daniestevez/gr-satellites#332] [Satellite teams]
# IS1 Introduction
The INSPIRESat-1 is a to-be-launched nanosatellite from the International Space Program In Research and Education (INSPIRE)  Educate to promote cultural exchanges through international collaborative space science and spacecraft engineering projects.
It consists of 2 Payloads:
1. Compact Ionospheric Probe (CIP): To improve the understanding of Ionosphere dynamics through observations of ion temperature, composition, density, and velocity by Compact Ionosphere Probe (CIP).
2. Dual Aperture X-ray Solar Spectrometer (DAXSS): To improve our understanding of the sun’s coronal heating processes by measuring the Soft X-Ray spectrum of the sun.

## Telemetry Beacon
We have a open beacon definition for the community, 

## Transmitter Description
|Band          |70cm                |
|---------------|---------------------|
|Mfg./Model    |SpaceQuest TRX-U     | 
|Frequency     |437.500 MHz          | 
|Modulation    |GFSK                 | 
|Baudrate  |9.6k                   | 
|Framing       |AX.25/G3RUH (Scrambler and GFSK decoder are swapped in order)          | 
|Antenna       |Low-gain λ/4 monopole| 
|Transmit power |1.75W (32.43 dBm)    | 

The flowgraph used for reception using USRP blocks, [here](https://github.com/Dhruva-Ananth/IS1_Comissioning/blob/main/IIST_IS1_UHF_SDR_v1_rxonly.grc)

Other than the flowgraph other external dependencies for the operations can be used and are as follows:
1. [gr-gpredict-doppler](https://github.com/wnagele/gr-gpredict-doppler)
2. [gr-ais](https://github.com/bistromath/gr-ais)
3. [gr-kiss](https://github.com/daniestevez/gr-kiss)

## Beacon Packet Definition

Once we receive the PDUs after HDLC de-framing (CRC, Preamble, byte ordering) the following packet structure.

### Framing
All data sent by the CUTE spacecraft over the 70cm (437.250 MHz) link are contained within AX.25 unnumbered information (UI) frames:

|    |Source Call Sign|Destination Call Sign|Control|PDU   |
|------|----------------|---------------------|-------|------|
|Size   |7 Bytes         |7 bytes              |1 byte |1 byte|
|Value  |'BCT   0'        |'IS-1  p'             |0x03   |0xF0  |

The control field is always set to 0x03, which identifies the AX.25 frames as unnumbered frames. The PID field is always set to 0xF0, which indicates that no layer 3 protocol is used.

### Source and Destination Call Signs

| Byte|Source Charecter|Source Hex|Destination Charecter|Destination Hex|
|------|----------------|---------------------|-------|------|
|1   |'B'        |0x84           |'I'       |0x92|
|2   |'C'        |0x86           |'S'      |0xa6|
|3   |'T'        |0xA8           |'-'      |0x5a|
|4   |' '         |0x40           |'1'      |0x62|
|5   |' '         |0x40           |' '       |0x40|
|6   |' '         |0x40           |' '       |0x40|
|7   |'0'        |0x60           |'p'      |0x60|

Note that the 'hex' column shows the numeric value corresponding to the ASCII character displayed in the 'char' column before it is a bit shifted to obtain a said character.

6 Byte header for the CCSDS packet.

| Element | Description       | Data Type |
|----------|------------------|------------|
|VERSION|CCSDS Version  |U3              |
|TYPE       |CCSDS Type      |U1              |
|Secondary Header Flag|   |U1              |
|Packet APID  |                   |U11            |
|Sequnce Flags |                |U2              |
|SRC Sequence Control |   |U14            |
|Packet Length|                  |U16           |

### Beaconing
The 70cm transceiver is configured to beacon at a rate of once every 10 seconds, though this rate may be changed post-commissioning.

Each beacon consists of a single state of health (SOH) packet, which fits in a single AX.25 UI frame: these packets contain 256 bytes of data (including 6 bytes of CCSDS primary header).

More info is in this [Repository](https://github.com/Dhruva-Ananth/INSPIRESat_1_Beacon_Info), for the extended beacon definition.

## To-Do
1. Not very sure on how to go ahead to add this decoder info into the gr-satellites systems, hope someone can help guide us through that!
2. In the decoding operation we have the G3RUH scrambler and NRZI, reversed in order any suggestions on how can we accommodate that in the decode chain?

### More resources
GRC scripts for PDU generation are available at this [repo](https://github.com/Dhruva-Ananth/INSPIRESat_1_Beacon_Info/blob/main/IIST_IS1_UHF_SDR_v1_rxonly.grc), for some reference !

Thank you very much!

**Comment by daniestevez** [comment]
Hi! Thanks for the information. I'm first moving this to a [discussion](https://github.com/daniestevez/gr-satellites/discussions), since now we handle threads devoted to support to satellite teams as discussions.

**Comment by daniestevez** [comment]
My understanding is that the modulation and coding used by INSPIRESat-1 is standard 9k6 FSK AX.25 G3RUH packet radio. Regarding your question:

> In the decoding operation we have the G3RUH scrambler and NRZI, reversed in order any suggestions on how can we accommodate that in the decode chain?

These two operations commute, so the order in which you perform them doesn't matter.

Therefore, I think that the following SatYAML file will be enough to obtain frames from your satellite (this is using the temporary NORAD from SatNOGS DB):

```
name: INSPIRESat-1
norad: 99400
data:
  &tlm Telemetry:
    telemetry: ax25
transmitters:
  9k6 FSK downlink:
    frequency: 437.500e+6
    modulation: FSK
    baudrate: 9600
    framing: AX.25 G3RUH
    data:
    - *tlm
 ```

Can you test if this SatYAML file works with your satellite?

The next step would be to write a telemetry parser for your frames using [construct](https://construct.readthedocs.io/). Your frame structure seems very similar to [that of CUTE](https://github.com/daniestevez/gr-satellites/discussions/312), perhaps due to the use of the same SpaceQuest TRX-U transceiver (and in fact you mention CUTE in your message). Code for CUTE is [already available here](https://github.com/daniestevez/gr-satellites/blob/master/python/telemetry/cute_70cm.py). Can you take a look at this code and the documentation for CUTE and check if what you need is exactly the same or there are some differences? Depending on the differences, we may want to extract the common part to a telemetry parser for the TRX-U and then have the differences separately. Please give me some feedback on this aspect so that we can determine how best to proceed.

**Comment by c-decicco** [comment]
Hi @daniestevez,

> Your frame structure seems very similar to https://github.com/daniestevez/gr-satellites/discussions/312, perhaps due to the use of the same SpaceQuest TRX-U transceiver (and in fact you mention CUTE in your message).

Good eye! INSPIRESat-1 does indeed have some similarities with CUTE, as it was developed by IIST, NCU, and NTU in partnership with LASP, and as such uses some LASP code in its FSW.

> I'm first moving this to a [discussion](https://github.com/daniestevez/gr-satellites/discussions), since now we handle threads devoted to support to satellite teams as discussions.

Sorry about that—I have been working with @Dhruva-Ananth to get support for INSPIRESat-1 into gr-satellites (as well as SatNOGS): he is following some internal written guidance I put together when I did the same for CUTE that is now a little stale—I need to update this to direct our engineers and collaborators to open discussions and not issue reports for new teams. (This is also why CUTE appears in the original post—a copy/paste error from the original CUTE gr-satellites discussion).

Along these lines—I have a parser that is, I think, just about ready to be merged that I have pushed to [lasp/gr-satellites](https://github.com/lasp/gr-satellites/tree/master) fork as commit [`ab7848a5`](https://github.com/lasp/gr-satellites/commit/ab7848a50fae166e94e23b35a4b06a528d7fea14).

Outstanding issues are that I think a conversion needs to be added for `bat_vcell`, and this parser may be amenable to refactoring to share some code with CUTE (perhaps the CCSDS portions should be refactored out to a separate file)? It is worth noting that the secondary headers are slightly different (the `sh_fine` time field in the IS-1 packet is 2 bytes long, unlike the similar sub-seconds field for CUTE which is 1 byte long), as well as the primary header (CUTE uses the uppermost bit of the APID to indicate whether the data is real time or stored—I am not sure if INSPIRESat-1 does this as well or not).

> Can you test if this SatYAML file works with your satellite?

I think this SatYAML file is correct in every regard except a NORAD ID has since been assigned, [51657](https://gitlab.com/librespacefoundation/satnogs-ops/-/issues/211#note_841464287).

@Dhruva-Ananth, could you please test @daniestevez's SatYAML file (with the corrected NORAD ID)?

**Comment by daniestevez** [comment]
Ni Nicholas,

Thanks! I've taken a look at [ab7848a5](https://github.com/lasp/gr-satellites/commit/ab7848a50fae166e94e23b35a4b06a528d7fea14) and it looks good. I'd rather `raise NotImplementedError` or something similar rather than returning zero in `PolynomialAdapter._encode()`, since returning zero could make using this inadvertently go unnoticed.

I think it could be good to separate the CCSDS portions into a separate file, but as you mention, there is always the risk of making the code too complicated to handle small differences. Time codes are often project-specific, for instance, so I wouldn't try to make general code for that. In case it seems too complicated, I think it's OK not to share this code. After all, it's just going to be a dozen line listing the fields in the CCSDS packet headers, so it's not much of a problem if this is repeated with some small variations.

**Comment by c-decicco** [comment]
Hi Daniel,

> I'd rather raise NotImplementedError or something similar rather than returning zero in PolynomialAdapter.\_encode(), since returning zero could make using this inadvertently go unnoticed.

Ah, thanks for reminding me of that--I would have forgotten otherwise. I have modified `PolynomialAdapter` in commit [`457c9970`](https://github.com/lasp/gr-satellites/commit/457c99705eaf8c605ee2a4552860984304a3142a) to raise this exception.

> I think it could be good to separate the CCSDS portions into a separate file, but as you mention, there is always the risk of making the code too complicated to handle small differences.

I agree that the CCSDS primary header, at least, could be refactored. I've since realized that there was a CCSDS primary header definition already available in [`python/ccsds/space_packet.py`](https://github.com/daniestevez/gr-satellites/blob/master/python/ccsds/space_packet.py) that I could have used for this (as well as for CUTE, the high bit of CUTE's APID notwithstanding), so likwise in commit `457c9970`, I changed the code to use this CCSDS primary header definition. (I can do the same for the CUTE parser, but I think you would agree that this should be done in a separate commit to keep things more atomic.)

Other changes I made in this commit:
* The parser structure is now called `inspiresat_1`, to be consistent with the naming scheme of the other parsers (i.e., where the parser struct is named the same as the file).
* The parser is now loaded in `__init__.py`.

By the way, for future LASP CubeSat missions that are using the same codebase, the secondary header could at least be refactored into a separate file: missions coming up for which I _think_ this would apply would be [CTIM](https://techport.nasa.gov/view/94898) and [SPRITE](https://space.skyrocket.de/doc_sdat/sprite.htm) (as well as some others that are futher off whose names I can't quite recall...). I believe CTIM will be launching later this year. SPRITE is further off. I would have to talk with our CubeSat FSW lead to know how similar the telemetry is between these missions. I think that refactoring would best be done in future when support for those CubeSats is added to gr-satellites.

It is worth noting that only the CUTE sw\_stat packet is generated by LASP software: the other packets are generated by the BCT bus used on that CubeSat. This is why the APID in the primary header and the time format in the secondary header are different between CUTE and INSPIRESat-1. Along these lines, [CIRBE](https://lasp.colorado.edu/home/cirbe/) is another LASP CubeSat which is also slated to launch later this year (actually, on the same rocket as CTIM) which also uses a BCT bus (though of one generation newer), and so should have more similarity to CUTE than to INSPIRESat-1/CTIM/SPRITE. CSIM is another LASP CubeSat currently on orbit which uses a BCT bus (of the same generation as CUTE) and should therefore be very similar to CUTE. There is currently a SatNOGS decoder for CSIM, though not a gr-satellites one. So really, I suppose CSIM, CIRBE, CUTE, and any other BCT XB1-based CubeSats could share the same CCSDS secondary header definition and time adapter.

Anyway, I also went ahead and pushed a second commit to lasp/gr-satellites with the SatYAML file added, [`40dc6d74`](https://github.com/lasp/gr-satellites/commit/40dc6d748b62038e53444fae5fd617328348b116).

I am attaching a GNU Radio 3.9 flowgraph to this comment as well as a 48 kHz audio file containing INSPIRESat-1 beacons (in raw 32-bit float format) that can be used to test the satellite YAML file as well as the telemetry decoder. The audio file is a cut-down version of SatNOGS observation [#5478149](https://network.satnogs.org/observations/5478149/). The .zip file containing these two files is available here: [inspiresat.zip](https://github.com/daniestevez/gr-satellites/files/8092559/inspiresat.zip)

Anyway, please let me know if you have any other changes you'd like to see made, or if you'd like me to go ahead and open a pull request.

Thanks,

Nick

**Reply by daniestevez** [reply]
Hi Nick,

i think everything seems in order for a pull request.

Take note that I've recently merged #335 from @janvgils, which adds a SatYAML file for INSPIRESat-1. Only the `telemetry:` line would need to be changed.

Regarding refactoring for future missions, I agree that this is best left to the future, when it's clearer what similarities and differences these satellites have.

Thanks for the contribution!

**Reply by c-decicco** [reply]
Hi @daniestevez,

I went ahead and rebased my commits onto `main`, and squashed my commit which adds the .yml file into the commit which adds the parser, which when combined with @janvgils' commit, means that the new commit changes only the `telemetry:` line.

I then went ahead and opened pull request #338.

Thanks!

Nick
